src/rgw/rgw_rest_s3.cc

```diff
@@ -4262,6 +4262,122 @@ rgw::auth::s3::LocalEngine::authenticate(
   return result_t::grant(std::move(apl), completer_factory(k.key));
 }
 
+/* OPAEngine */
+rgw::auth::Engine::result_t
+rgw::auth::s3::OPAEngine::authenticate(
+  const boost::string_view& _access_key_id,
+  const boost::string_view& signature,
+  const string_to_sign_t& string_to_sign,
+  const signature_factory_t& signature_factory,
+  const completer_factory_t& completer_factory,
+  const req_state* const s) const
+{
+  /* get the user info */
+  RGWUserInfo user_info;
+  const std::string access_key_id = _access_key_id.to_string();
+  if (rgw_get_user_info_by_access_key(store, access_key_id, user_info) < 0) {
+      ldout(cct, 5) << "error reading user info, uid=" << access_key_id
+              << " can't authenticate" << dendl;
+      return result_t::deny(-ERR_INVALID_ACCESS_KEY);
+  }
+
+  /* ask OPA for a decision */
+
+  /* get OPA url */
+  char * opaUrl = getenv("OPA_URL");
+  if (opaUrl == NULL) {
+    dout(10) << "OPA_URL env variable not set" << dendl;
+    return result_t::deny(-ERR_INVALID_REQUEST);
+  }
+  dout(10) << "OPA URL= " << std::string(opaUrl) << dendl;
+
+  /* get authentication token for OPA */
+  char * token = getenv("OPA_AUTHENTICATION_TOKEN");
```

Comment:
How are the lifetimes of opa tokens and how are they obtained? They should be either set statically via ceph.conf or obtained from the opa server and then used

Reply:
OPA uses Bearer tokens to authenticate client requests. The token will be passed through ceph.conf.